controllers/llamaCPP.cc

```diff
@@ -654,7 +663,6 @@ bool llamaCPP::loadModelImpl(std::shared_ptr<Json::Value> jsonBody) {
 
   queue = new trantor::ConcurrentTaskQueue(llama.params.n_parallel, "llamaCPP");
 
-  llama.model_loaded_external = true;
```

Comment:
The hang issue is due to this, @BachDao. There is a run loop in backgroundTask for updating results  Without `update_slots` being executed, `next_result` would wait until then (another run loop and mutex).

Reply:
Why the background task need to check that flag?
I assume that before you submit any task to background worker, everything should be initialized, or I missing smt ?

Reply:
I don't like this check flag either, but the `warmupModel` should not affect the `loadModel` request. That's why the flag is flipped before. There is another clue: when running the model warm-up process, there could also be an unload model request, and in this case, the flag check works (to prevent slot update). However, I'm not a fan of this logic; will try to refactor it somehow soon.

Reply:
Not sure if this is the appropriate place to discuss but I put a few lines below:

- As my understanding, `warmupModel` is called only once for each time model is loaded. If that correct, should both functions be placed on the same thread and run sequentially without synchronization ?
- Regarding `model_loaded_external`, it seems that the flag is used both to denote a successful model load and the continuation of the background task, should it be separated ?
https://github.com/janhq/nitro/blob/2242b7da8ef741e414f70a6be15128a7318186e6/controllers/llamaCPP.cc#L666-L669
- Should the task queue be wrapped behind a simple interface that allows submission, polling, and blocking-wait for results? The current implementation scattered in the controller class makes it a bit hard to follow. Also, having a unified interface provides flexibility for adapting to other use cases if necessary.
-  Will the following block the event loop? Should we consider submitting the continuation and freeing the I/O thread for another request?
https://github.com/janhq/nitro/blob/2242b7da8ef741e414f70a6be15128a7318186e6/controllers/llamaCPP.cc#L419-L421

Putting all these detail aside, it's still not clear to me what the use-cases of Nitro are:

- Is it intended to be used as an embedded server (single-user)? In that case, would a single I/O thread be sufficient? With current implementation of Drogon, for each I/O thread, there is separate event loop that keeps polling the Acceptor's socket, it's too much for single connection.
- Otherwise, if we intend to server multiple client,  the current implementation ties the lifetime of `llama_server_context` to the `controller` object, which has only single-instance for the entire app lifetime.
https://github.com/janhq/nitro/blob/2242b7da8ef741e414f70a6be15128a7318186e6/controllers/llamaCPP.h#L2250-L2251
It mean single inference context will be shared among all client, seem not appropriate. Imagine a scenario where the second client attempts to load a model with specific config but loses in the race, and the configuration comes from the first client.

Regardless, you guys done an excellent job making local inference easy for everyone. Looking forward to your future releases.

Reply:
- `warmupModel` is a background task...
- Agreed, as above. Should decouple the states...
- Sorry, there are many tasks mentioned here....
**_===> For all the points above, I will try to find sometime and come back with concrete code, it will be easier for discussion._**
-  This is not blocking the event loop; you can see it runs in another thread **_==> agree, the task itself running on background thread, but `next_result(task_id)`, which called on i/o thread can be blocking_wait if task not yet completed._**
- ... "Current implementation of Drogon" Are you referring to the SSE impl? **_==> Drogon, the HttpServer framework that used by Nitro._**